modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaDescriptor.java

```diff
@@ -84,71 +82,118 @@ public class SchemaDescriptor {
      * @param keyCols Key columns.
      * @param valCols Value columns.
      */
+    @TestOnly
     public SchemaDescriptor(int ver, Column[] keyCols, Column[] valCols) {
-        this(ver, keyCols, null, valCols);
+        this(
+                ver,
+                mergeColumns(keyCols, valCols),
+                Arrays.stream(keyCols).map(Column::name).collect(Collectors.toList()),
+                null
+        );
     }
 
-    /**
-     * Constructor.
-     *
-     * @param ver     Schema version.
-     * @param keyCols Key columns.
-     * @param colocationCols Colocation column names.
-     * @param valCols Value columns.
-     */
-    public SchemaDescriptor(int ver, Column[] keyCols, String @Nullable[] colocationCols, Column[] valCols) {
-        assert keyCols.length > 0 : "No key columns are configured.";
+    /** Constructor. */
+    public SchemaDescriptor(
+            int ver,
+            List<Column> columns,
+            List<String> keyColumns,
+            @Nullable List<String> colocationColumns
+    ) {
+        assert !nullOrEmpty(columns) : "Schema should have at least one column";
+
+        Map<String, Column> columnsByName = new HashMap<>();
+        List<Column> orderedColumns = new ArrayList<>(columns.size());
+
+        Object2IntMap<String> columnNameToPositionInKey = new Object2IntOpenHashMap<>();
+        int idx = 0;
+        for (String name : keyColumns) {
+            assert !columnNameToPositionInKey.containsKey(name)
+                    : "Key column must not have duplicates: " + name;
+
+            columnNameToPositionInKey.put(name, idx++);
+        }
```

Comment:
This can be an utility method.

Reply:
the goal of this patch is to remove code rather than adding new one, so I tried not add new lines until it's absolutely required in order to proceed. There was no validation before, but after refactoring this object become more sensible to unspoken invariants like "no duplicate columns". In order to find and all the places that violates such invariant I've added bunch of assertions.

I think, the proper fix of validation should be done under separate issue (https://issues.apache.org/jira/browse/IGNITE-21733).